### DISCUSSION TITLE
Integration in Unity

### DISCUSSION THREAD

**Discussion by DavidPolastik** [erew123/alltalk_tts#411] [Q&A]
Hi there,

I'm currently working with the TTS streaming API in Unity (located at http://127.0.0.1:7851/api/tts-generate-streaming) for generating real-time TTS audio. However, I'm encountering an issue where I get a 404 Not Found error when trying to stream the audio.

Generation in start_alltalk working correctly

What I'm trying to do:
I make a POST request to the API with the necessary parameters (text, voice, language, and output file).
The API responds with a JSON containing the output file path (e.g., "output_file_path": "stream_output.wav").
I then try to stream the audio using the URL http://127.0.0.1:7851/stream_output.wav.
Problem:
When I try to access the URL for the generated file (http://127.0.0.1:7851/stream_output.wav), I get a 404 Not Found error.

Code:
public class TTSStreaming : MonoBehaviour
{
    private string apiUrl = "http://127.0.0.1:7851/api/tts-generate-streaming"; // URL pro generování TTS
    public string textInput = "Here is some text to be converted into speech."; // Text pro převod na řeč
    public string voice = "female_01.wav"; // Jméno hlasu
    public string language = "en"; // Jazyk pro TTS
    public AudioSource audioSource; // Přiřaďte komponentu AudioSource na GameObject v Unity

    void Start()
    {
        StartCoroutine(GenerateAndStreamTTS(textInput, voice, language));
    }

    IEnumerator GenerateAndStreamTTS(string text, string voice, string language)
    {
        // Parametry pro požadavek na API (metoda POST)
        WWWForm form = new WWWForm();
        form.AddField("text", text);
        form.AddField("voice", voice);
        form.AddField("language", language);
        form.AddField("output_file", "stream_output.wav"); // Název souboru (v tomto případě není potřeba, ale server to vyžaduje)

        using (UnityWebRequest www = UnityWebRequest.Post(apiUrl, form))
        {
            // Poslání požadavku na API
            yield return www.SendWebRequest();

            if (www.result == UnityWebRequest.Result.Success)
            {
                // Získání odpovědi z API (ve formátu JSON)
                string responseText = www.downloadHandler.text;
                Debug.Log("API odpověď (text): " + responseText);

                // Deserializace JSON odpovědi pro získání cesty k souboru
                var jsonResponse = JsonUtility.FromJson<ResponseData>(responseText);
                string outputFilePath = jsonResponse.output_file_path;

                // Ověřte, zda soubor není prázdný
                if (!string.IsNullOrEmpty(outputFilePath))
                {
                    // Vytvoření URL pro streamování (streaming URL)
                    string streamingUrl = $"http://127.0.0.1:7851/{outputFilePath}";

                    // Debugování URL pro streamování
                    Debug.Log("URL pro streamování: " + streamingUrl);

                    // Pokusíme se streamovat audio
                    using (UnityWebRequest audioRequest = UnityWebRequest.Get(streamingUrl))
                    {
                        audioRequest.SendWebRequest();

                        // Čteme data jako audio stream
                        while (!audioRequest.isDone)
                        {
                            yield return null; // Čekáme na dokončení požadavku
                        }

                        if (audioRequest.result == UnityWebRequest.Result.Success)
                        {
                            byte[] audioData = audioRequest.downloadHandler.data;

                            if (audioData.Length > 0)
                            {
                                try
                                {
                                    // Převedení audio streamu na AudioClip
                                    AudioClip audioClip = WavUtility.ToAudioClip(audioData);
                                    audioSource.clip = audioClip;
                                    audioSource.Play();
                                }
                                catch (System.Exception ex)
                                {
                                    Debug.LogError("Chyba při konverzi WAV dat: " + ex.Message);
                                }
                            }
                            else
                            {
                                Debug.LogError("Přijatá audio data jsou prázdná.");
                            }
                        }
                        else
                        {
                            Debug.LogError("Chyba při streamování audio: " + audioRequest.error);
                        }
                    }
                }
                else
                {
                    Debug.LogError("Chyba: Soubor neexistuje.");
                }
            }
            else
            {
                Debug.LogError("Chyba při posílání požadavku: " + www.error);
            }
        }
    }

    [System.Serializable]
    public class ResponseData
    {
        public string output_file_path; 
    }
}
`

**Comment by erew123** [comment]
@DavidPolastik Im not sure if it does or does not work in Unity or what any limitations of unity may be. I am also not a unity coder and I am super busy with other code and life things at the moment. However, thrown my current knowledge into an AI to get answers to questions for you. I will just say I personally do not know if Unity supports PCM streams of this type, however, hopefully everything I am about to give you will give you useful information to research or test.

Also, I assume you are using XTTS as the engine, as that is the ONLY one that supports streaming currently.

There are 3x sections below:

🟥 How the XTTS streaming audio is generated
🟥 Possible code & Debugging suggestions
🟥 3rd Party Options that can be imported into Unity to play audio

Hopefully there are some options for you below. Good luck!

---

# 🟥 How the XTTS streaming audio is generated:
```
        if streaming:
            # Streaming-specific operations
            file_chunks = []
            wav_buf = io.BytesIO()
            with wave.open(wav_buf, "wb") as vfout:
                vfout.setnchannels(1)
                vfout.setsampwidth(2)
                vfout.setframerate(24000)
                vfout.writeframes(b"")
            wav_buf.seek(0)
            yield wav_buf.read()

            for i, chunk in enumerate(output):
                file_chunks.append(chunk)
                if isinstance(chunk, list):
                    chunk = torch.cat(chunk, dim=0)
                chunk = chunk.clone().detach().cpu().numpy()
                chunk = chunk[None, : int(chunk.shape[0])]
                chunk = np.clip(chunk, -1, 1)
                chunk = (chunk * 32767).astype(np.int16)
                yield chunk.tobytes()
```

This is generating audio as a **16-bit PCM stream** with the following specifications:

- **Audio Format**: Linear PCM (Pulse-Code Modulation)
- **Bit Depth**: 16-bit
- **Channels**: Mono (1 channel)
- **Sample Rate**: 24,000 Hz
- **Data Encoding**: Signed integers (`int16`) with a range of values between `-32768` and `32767`
- **Data Delivery**: Chunked in small segments during streaming

The audio stream is being sent with a MIME type of `audio/wav`, though the actual data being streamed is raw PCM encapsulated within a basic WAV header.

### Debugging Notes:
- **Streaming Behavior**: The data is streamed in chunks, which may require client-side buffering or reconstruction to ensure smooth playback.
- **Data Clipping**: The PCM data is pre-clipped to ensure it stays within valid ranges for 16-bit audio.
- **Output Format**: PCM data is provided as raw bytes, where each audio sample is represented as two bytes (16 bits).

### Suggestions for Debugging:
- Use tools like `ffplay` or `sox` to test the stream.
- Inspect the initial bytes of the stream (WAV header) to ensure correctness.
- Confirm the client (Unity) can handle and interpret the 16-bit PCM mono stream at 24000 Hz.

---

# 🟥 Possible code & Debugging suggestions
Re your code above - the issue seems to be a misunderstanding of how the streaming and file access processes are implemented in the API and Unity code. Potential issues and solutions:

### 1. **Mismatch Between Streaming and File Download**:
   - The endpoint `/api/tts-generate-streaming` is designed to provide a streaming audio response, not a static file to be downloaded from a separate URL like `http://127.0.0.1:7851/stream_output.wav`.
   - The Unity client code appears to be attempting to retrieve the file `stream_output.wav` as a static file, but this file is not hosted on the server in the way it expects. The server is not configured to serve static files.

   **Solution**:
   - If you want to stream audio directly from the `/api/tts-generate-streaming` endpoint, the Unity client should connect to that endpoint and process the streamed data in real-time.
   - Update the Unity code to handle the stream directly from the API without expecting a separate file.

---

### 2. **Unity Support for PCM Streams**:
   - Unity's audio system (`AudioClip` and `AudioSource`) supports PCM data, but it needs to be converted into the format Unity expects (`AudioClip`).
   - The `WavUtility.ToAudioClip` function may not work directly with PCM chunks because it is expecting complete WAV-formatted data.

   **Solution**:
   - Modify the Unity code to handle raw PCM streams. The streamed PCM data from the server needs to be reconstructed into chunks that Unity can process. You might need to write a custom function to create an `AudioClip` from PCM data.

---

### 3. **Streaming Endpoint Configuration**:
   - The server is correctly set up to stream PCM data, but the client (Unity) is not handling this correctly.
   - Ensure that the server sends data in small chunks with proper headers (`audio/wav`) as it does in your code.

   **Solution**:
   - Test the streaming endpoint with tools like `cURL` or a custom Python script to confirm the server is sending the expected data.
   - Update Unity's `UnityWebRequest` code to handle streaming more effectively. Unity supports downloading and reading data in chunks using `DownloadHandlerScript`.

---

### Updated Unity Approach for Streaming
Here’s a Unity script that directly processes the streamed audio from the server:

```csharp
using UnityEngine;
using UnityEngine.Networking;
using System.Collections;

public class TTSStreaming : MonoBehaviour
{
    private string streamingUrl = "http://127.0.0.1:7851/api/tts-generate-streaming"; // Streaming endpoint
    public AudioSource audioSource;

    void Start()
    {
        StartCoroutine(StreamAudio());
    }

    IEnumerator StreamAudio()
    {
        using (UnityWebRequest request = UnityWebRequest.Get(streamingUrl))
        {
            request.downloadHandler = new DownloadHandlerBuffer(); // Buffer for streamed data
            request.SendWebRequest();

            while (!request.isDone)
            {
                yield return null; // Wait for the response
            }

            if (request.result == UnityWebRequest.Result.Success)
            {
                byte[] audioData = request.downloadHandler.data;

                if (audioData != null && audioData.Length > 0)
                {
                    // Convert PCM data to AudioClip
                    AudioClip audioClip = ConvertPCMToAudioClip(audioData);
                    if (audioClip != null)
                    {
                        audioSource.clip = audioClip;
                        audioSource.Play();
                    }
                    else
                    {
                        Debug.LogError("Failed to convert PCM data to AudioClip.");
                    }
                }
                else
                {
                    Debug.LogError("No audio data received.");
                }
            }
            else
            {
                Debug.LogError($"Error: {request.error}");
            }
        }
    }

    private AudioClip ConvertPCMToAudioClip(byte[] pcmData)
    {
        // Assuming 16-bit PCM mono data at 24000 Hz as per your server's configuration
        int sampleRate = 24000;
        int totalSamples = pcmData.Length / 2; // 16-bit PCM
        float[] floatSamples = new float[totalSamples];

        for (int i = 0; i < totalSamples; i++)
        {
            short sample = System.BitConverter.ToInt16(pcmData, i * 2);
            floatSamples[i] = sample / 32768.0f; // Normalize to range [-1.0, 1.0]
        }

        AudioClip audioClip = AudioClip.Create("TTS_Audio", totalSamples, 1, sampleRate, false);
        audioClip.SetData(floatSamples, 0);
        return audioClip;
    }
}
```

---

### 4. **File-Based Solution as a Backup - AKA don't stream**:
If you still want to support file-based playback:
   - Serve the generated file as a static resource. For example, configure your server to host files from a directory (`output_directory`) so the client can access them via a URL.

---

### Final Notes:
- The server and Unity must agree on the format of the streamed data.
- Ensure proper error handling and debugging on both ends to identify any additional issues.
- Use tools like `Postman` or a custom script to validate the server's streaming behavior independently before integrating with Unity.

https://docs.unity3d.com/ScriptReference/AudioClip.html

https://docs.unity3d.com/ScriptReference/AudioSource-clip.html

---

# 🟥 3rd Party Options that can be imported into Unity to play audio

### **1. FMOD (Feature-Rich Audio Middleware)**

- **What It Is**:
  FMOD is a professional-grade audio middleware that integrates seamlessly with Unity. It supports advanced features like streaming audio, 3D sound, and dynamic audio playback.
  
- **Why It's Useful**:
  FMOD can handle audio streams, including custom formats like PCM, and is designed for real-time processing and playback.

- **How to Use**:
  - Download the FMOD Unity integration from the FMOD website or Unity Asset Store.
  - Configure FMOD to read and process PCM streams from your TTS server.
  - FMOD has APIs for handling raw audio streams and custom audio input sources.

- **Documentation**: [FMOD Unity Integration](https://fmod.com/resources/documentation)

---

### **2. NAudio (C# Audio Library)**

- **What It Is**:
  NAudio is a powerful C# audio library that supports many audio formats and operations, including raw PCM playback. While it is not native to Unity, it can be integrated with Unity scripts.

- **Why It's Useful**:
  NAudio can directly process PCM streams and offers robust audio handling capabilities that may surpass Unity's built-in `AudioSource`.

- **How to Use**:
  - Import the NAudio library into your Unity project (you may need to convert it to a .NET Standard DLL).
  - Use NAudio’s `WaveStream` or `BufferedWaveProvider` to handle the PCM stream from your TTS engine.
  - Pass the processed audio data into Unity's `AudioSource` for playback.

- **Documentation**: [NAudio GitHub](https://github.com/naudio/NAudio)

---

### **3. UniWebView or Custom HTTP Stream Plugins**

- **What It Is**:
  If your TTS server provides audio streams in a common format (e.g., WAV, MP3), you can use a plugin like UniWebView to stream and play back audio directly.

- **Why It's Useful**:
  Plugins like UniWebView simplify web-based audio handling and can provide smoother playback than Unity's native `UnityWebRequest` for certain use cases.

- **How to Use**:
  - Install UniWebView from the Unity Asset Store.
  - Stream audio from your server and pass it to the Unity `AudioSource` for playback.

- **Documentation**: [UniWebView Official Site](https://www.uniwebview.com/)

---

### **4. BASS Audio Library**

- **What It Is**:
  BASS is a lightweight audio library for playing and managing audio streams. It is popular in game development due to its low overhead and robust streaming support.

- **Why It's Useful**:
  BASS supports raw PCM data and streaming from various sources, including HTTP streams. It can handle complex audio tasks outside Unity's native capabilities.

- **How to Use**:
  - Download the BASS library and its Unity wrapper from [Un4seen](https://www.un4seen.com/).
  - Use the BASS library to stream and decode audio, and pass it to Unity's `AudioSource`.

- **Documentation**: [BASS Documentation](https://www.un4seen.com/doc/)

---

### **5. Custom Audio Plugins (DSPs)**

- **What It Is**:
  Unity allows the creation of custom audio plugins using its DSP (Digital Signal Processing) framework. This can be used to handle raw audio streams directly.

- **Why It's Useful**:
  If your TTS system produces PCM streams in a specific way, a custom Unity DSP plugin can process the data in real-time, bypassing Unity's native limitations.

- **How to Use**:
  - Write a Unity DSP plugin in C++ or C#.
  - Use the plugin to receive PCM data and stream it into Unity’s audio pipeline.

- **Documentation**: [Unity Custom Audio Plugins](https://docs.unity3d.com/Manual/AudioMixerCustomEffect.html)

---

### **6. Other Streaming Libraries**

- **Photon Voice**: Primarily for voice chat, but it can handle audio streams.
- **Audiokinetic Wwise**: Similar to FMOD, it’s a robust audio middleware that supports streaming.

---

### Recommendations for Your Use Case
- Start with **FMOD** or **BASS**, as they have built-in streaming capabilities and work well with Unity.
- If you need lightweight integration, consider **NAudio** or a Unity-native plugin.
- Test and debug using simpler tools like **UniWebView** or **custom HTTP clients** if you're unsure about PCM compatibility.
